src/chain/chaininterface.rs

```diff
@@ -54,44 +60,82 @@ pub enum ConfirmationTarget {
 pub struct ChainWatchInterfaceUtil {
 	watched: Mutex<(Vec<Script>, Vec<(Sha256dHash, u32)>, bool)>, //TODO: Something clever to optimize this
 	listeners: Mutex<Vec<Weak<ChainListener>>>,
+	reentered: AtomicUsize
 }
 
-impl ChainWatchInterfaceUtil {
-	pub fn new() -> ChainWatchInterfaceUtil {
-		ChainWatchInterfaceUtil {
-			watched: Mutex::new((Vec::new(), Vec::new(), false)),
-			listeners: Mutex::new(Vec::new()),
-		}
-	}
-
-	pub fn install_watch_script(&self, spk: Script) {
+/// Register listener
+impl ChainWatchInterface for ChainWatchInterfaceUtil {
+	fn install_watch_script(&self, script_pub_key: Script) {
 		let mut watched = self.watched.lock().unwrap();
-		watched.0.push(Script::from(spk));
+		watched.0.push(Script::from(script_pub_key));
+		self.reentered.fetch_add(1, Ordering::Relaxed);
 	}
 
-	pub fn install_watch_outpoint(&self, outpoint: (Sha256dHash, u32)) {
+	fn install_watch_outpoint(&self, outpoint: (Sha256dHash, u32)) {
 		let mut watched = self.watched.lock().unwrap();
 		watched.1.push(outpoint);
+		self.reentered.fetch_add(1, Ordering::Relaxed);
 	}
 
-	pub fn watch_all_txn(&self) { //TODO: refcnt this?
+	fn watch_all_txn(&self) {
 		let mut watched = self.watched.lock().unwrap();
 		watched.2 = true;
+		self.reentered.fetch_add(1, Ordering::Relaxed);
 	}
 
-	pub fn register_listener(&self, listener: Weak<ChainListener>) {
+	fn register_listener(&self, listener: Weak<ChainListener>) {
 		let mut vec = self.listeners.lock().unwrap();
 		vec.push(listener);
 	}
+}
+
+impl ChainWatchInterfaceUtil {
+	pub fn new() -> ChainWatchInterfaceUtil {
+		ChainWatchInterfaceUtil {
+			watched: Mutex::new((Vec::new(), Vec::new(), false)),
+			listeners: Mutex::new(Vec::new()),
+			reentered: AtomicUsize::new(1)
+		}
+	}
 
+	/// notify listener that a block was connected
+	/// notification will repeat if notified listener register new listeners
+	pub fn block_connected(&self, block: &Block, height: u32) {
+		let mut watch = self.reentered.load(Ordering::Relaxed);
+		let mut last_seen = 0;
+		// re-scan if new watch added during previous scan
+		while last_seen != watch {
+			last_seen = watch;
+			let mut matched = Vec::new();
+			let mut matched_index = Vec::new();
+			{
+				let watched = self.watched.lock().unwrap();
+				for (index, transaction) in block.txdata.iter().enumerate() {
+					if self.does_match_tx_unguarded(transaction, &watched) {
+						matched.push(transaction);
+						matched_index.push(index as u32);
+					}
+				}
+			}
+			self.do_call_block_connected(&block.header, height, matched.as_slice(), matched_index.as_slice());
+			watch = self.reentered.load(Ordering::Relaxed);
+		}
+	}
+
+	/// notify listener that a block was disconnected
+	pub fn block_disconnected(&self, header: &BlockHeader) {
+		self.do_call_block_disconnected(header);
+	}
+
+	/// call listeners for connected blocks if they are still around. public only because of tests
```

Comment:
I think this should be public generally, and should probably do the reentry check inside of itself here. ie move the let mut watch = self.reentered.load... to above the calls, and then load it again at the end, returning true/false if the user needs to call it again. That will simplify block_connected a bit, and also enable people who havent done the full-Block download to use the API easier. Then maybe rename it from do_call to block_connected_checked and block_connected something like block_connected_with_filtering.